Add instructions for container metrics

MuriloDalRi opened the pull request:
Instructions on how to deploy the metrics exporter app to get container metrics such as CPU, RAM, disk usage data, app crashes, app requests and app response times.

MuriloDalRi committed:
Add instructions for container metrics

kentsanggds commented on source/manuals/monitor-paas-app-with-prometheus.html.md.erb:
```diff
@@ -36,6 +36,47 @@ You can create a Prometheus service within your PaaS space and allow it to bind
+	You can also set environment variables by amending the manifest file. We recommend that you use this method for optional environment variables that do not contain secret information. Refer to the [https://github.com/alphagov/paas-metric-exporter](https://github.com/alphagov/paas-metric-exporter) repository for more information.
+
+4. Start your app by running `cf start metric-exporter`.
+
```
Should also include information about how to secure the container metrics endpoint as teams may be concerned that the container metrics will expose information about the underlying infrastructure.

kentsanggds commented on source/manuals/monitor-paas-app-with-prometheus.html.md.erb:
```diff
@@ -36,6 +36,47 @@ You can create a Prometheus service within your PaaS space and allow it to bind
+
+4. Start your app by running `cf start metric-exporter`.
+
+You can now check your monitoring system to see if you are receiving metrics.
```
In order for the container metrics to be added as a target in prometheus it needs to be clear that they need to bind the prometheus service to the metric-exporter.

kentsanggds commented on source/manuals/monitor-paas-app-with-prometheus.html.md.erb:
```diff
@@ -36,6 +36,47 @@ You can create a Prometheus service within your PaaS space and allow it to bind
+
+### Troubleshooting
+
+If you are not receiving the metrics, check the [logs](/#logs) for the metrics exporter app. If you still need help, please contact us by emailing [gov-uk-paas-support@digital.cabinet-office.gov.uk](mailto:gov-uk-paas-support@digital.cabinet-office.gov.uk).
```
Is there a link missing here for `logs`?

kentsanggds committed:
Updated prometheus docs for container metrics

- also includes information about how to protect the endpoint with a route service.

soupdragon99 committed:
Tech writer edits

soupdragon99 commented:
@kentsanggds I have had a first pass at these changes.

soupdragon99 commented on source/manuals/monitor-paas-app-with-prometheus.html.md.erb:
```diff
@@ -39,21 +39,98 @@ You can create a Prometheus service within your PaaS space and allow it to bind
+
+[Cloud Foundry][] provides time-series data (metrics), for your PaaS apps.
+
+Use the metrics exporter app to send Cloud Foundry metrics data to a monitoring system of your choice to store and view it. You'll need to deploy the metrics exporter as an app on PaaS before you can use it.
```
Do we want to offer alternative solutions here? If so, we should give an example?

MuriloDalRi replied:
I don't think we have another way of doing this at the moment

soupdragon99 replied:
OK, in that case, I'd reword this sentence as we're saying there are other monitoring systems people could use,  but if we don't want people to do this I'd remove this comment to avoid any confusion about what options people have and by implication is supported.

soupdragon99 commented on source/manuals/monitor-paas-app-with-prometheus.html.md.erb:
```diff
@@ -39,21 +39,98 @@ You can create a Prometheus service within your PaaS space and allow it to bind
+
+### Set up the metrics exporter app
+
+Before you setting up the metrics exporter app, you'll need a live Cloud Foundry account assigned to the spaces you want to receive metrics on.
```
Is a link or any guidance we can link to that explains how to get a CF account?

MuriloDalRi replied:
I'll add a link

soupdragon99 commented on source/manuals/monitor-paas-app-with-prometheus.html.md.erb:
```diff
@@ -39,21 +39,98 @@ You can create a Prometheus service within your PaaS space and allow it to bind
+  - `cf bind-service <app-name> <service-instance-name>`
+
+### IP whitelist your app
+
```
We should have an intro here explaining why we have the section and why people should do this. It's probably self-explanatory for the target audience but it'll help those with less subject knowledge.

MuriloDalRi replied:
Will do

soupdragon99 commented on source/manuals/monitor-paas-app-with-prometheus.html.md.erb:
```diff
@@ -39,21 +39,98 @@ You can create a Prometheus service within your PaaS space and allow it to bind
+
+### Troubleshooting
+
+If you're not receiving metrics, check the [logs][] for the metrics exporter app or contact us at [gov-uk-paas-support@digital.cabinet-office.gov.uk][].
```
is there a Slack channel people can use?

philandstuff replied:
We should also be clear if it's re-tools or paas team supporting this. I don't think we should put the paas support email address on here unless we've asked them?

MuriloDalRi replied:
Should I remove that email and add our slack channel?

philandstuff replied:
yes, let's do that

philandstuff commented on source/manuals/monitor-paas-app-with-prometheus.html.md.erb:
```diff
@@ -39,21 +39,98 @@ You can create a Prometheus service within your PaaS space and allow it to bind
 
 Because changes to the first route will only be picked up within 5 minutes it could create gaps to metrics during route changes if you're using a zero downtime plugin such as [autopilot][].
 
+## Configure container metrics
```
I think this should maybe be a separate page?  Our user found it confusing mixing this content with the stuff above about the gds-prometheus service.

MuriloDalRi replied:
I can change that if you want, I also think it might make more sense

soupdragon99 replied:
I agree with doing that, especially if there's some UR suggesting it'd make more sense in a separate page, even if it's a sample of one.

philandstuff commented on source/manuals/monitor-paas-app-with-prometheus.html.md.erb:
```diff
@@ -39,21 +39,98 @@ You can create a Prometheus service within your PaaS space and allow it to bind
+
+1. Clone the [paas-metric-exporter][] GitHub repository.
+2. [Push the metrics exporter app](https://docs.cloud.service.gov.uk/#deployment-overview) to Cloud Foundry (without starting the app) by running
+ * `cf push --no-start metric-exporter`
```
The manifest in the metric exporter app repo has `no-route: true` so this command will launch the app without a route to scrape.  We need to work out how to do this properly.

MuriloDalRi replied:
I've added another step to map a route

philandstuff commented on source/manuals/monitor-paas-app-with-prometheus.html.md.erb:
```diff
@@ -39,21 +39,98 @@ You can create a Prometheus service within your PaaS space and allow it to bind
+1. Check you're generating Prometheus metrics at the metrics endpoint
+    - `https://<your-app-url>/metrics`
+5. Bind your app to the Prometheus service
+  - `cf bind-service <app-name> <service-instance-name>`
```
We say `<app-name>` here but `metric-exporter` above.  We should be consistent.

MuriloDalRi replied:
I'll change it to be consistent

philandstuff commented on source/manuals/monitor-paas-app-with-prometheus.html.md.erb:
```diff
@@ -39,21 +39,98 @@ You can create a Prometheus service within your PaaS space and allow it to bind
+
+4. Run `cf start metric-exporter` to start your app
+1. Check you're generating Prometheus metrics at the metrics endpoint
+    - `https://<your-app-url>/metrics`
```
what is `<your-app-url>`? This can be confusing because it's not the user's app (ie not their own code).

MuriloDalRi replied:
Will fix

MuriloDalRi committed:
Changes after PR review

philandstuff commented on source/manuals/monitor-paas-app-with-prometheus.html.md.erb:
```diff
@@ -94,12 +94,18 @@ To set up the metrics exporter app:
 4. Run `cf start metric-exporter` to start your app
 1. Check you're generating Prometheus metrics at the metrics endpoint
-    - `https://<your-app-url>/metrics`
+    - `https://metric-exporter.cloudapps.digital/metrics`
```
only one team can have this URL.  Maybe something like `https://<app-name>.cloudapps.digital/metrics`?

philandstuff commented on source/manuals/monitor-paas-app-with-prometheus.html.md.erb:
```diff
@@ -94,12 +94,18 @@ To set up the metrics exporter app:
-  - `cf bind-service <app-name> <service-instance-name>`
+  - `cf bind-service metric-exporter <service-instance-name>`
+6. Map a route to your app so Prometheus can scrape it
+  - `cf map-route metric-exporter cloudapps.digital --hostname metric-exporter`
```
this command will fail because the name `metric-exporter.cloudapps.digital` is taken I think. in any case, it'll only work once.

maybe `cf map-route metric-exporter cloudapps.digital --hostname <app-name>`?

philandstuff commented on source/manuals/monitor-paas-app-with-prometheus.html.md.erb:
```diff
@@ -94,12 +92,18 @@ To set up the metrics exporter app:
-  - `cf bind-service <app-name> <service-instance-name>`
+  - `cf bind-service metric-exporter <service-instance-name>`
+6. Map a route to your app so Prometheus can scrape it
+  - `cf map-route metric-exporter cloudapps.digital --hostname <app-name>`
```
these steps are in the wrong order - step one is "check the route is working", but step 6 is "set up the route".  We need to set the route before it'll work!

MuriloDalRi committed:
Adds new page for setting up paas metric exporter with prometheus

philandstuff commented on source/manuals/monitor-paas-app-with-prometheus.html.md.erb:
```diff
@@ -39,21 +39,25 @@ You can create a Prometheus service within your PaaS space and allow it to bind
 [update your app's manifest.yml]: https://docs.cloudfoundry.org/devguide/deploy-apps/manifest.html#services-block
+[Cloud Foundry]: https://docs.cloudfoundry.org/concepts/overview.html
+[paas-metric-exporter]: https://github.com/alphagov/paas-metric-exporter
+[gov-uk-paas-support@digital.cabinet-office.gov.uk]: mailto:gov-uk-paas-support@digital.cabinet-office.gov.uk
```
this definitely shouldn't be needed

philandstuff commented on source/manuals/monitor-paas-app-with-prometheus.html.md.erb:
```diff
@@ -39,21 +39,24 @@ You can create a Prometheus service within your PaaS space and allow it to bind
 
 Because changes to the first route will only be picked up within 5 minutes it could create gaps to metrics during route changes if you're using a zero downtime plugin such as [autopilot][].
 
-[#re-prometheus-support Slack channel]: https://gds.slack.com/messages/CAF5H4N4Q/
```
this reference is still needed

philandstuff replied:
i pushed a commit fixing the references, should be good to go now

philandstuff committed:
fix up references

make sure all used references exist, all stale references are removed,
and they're in a somewhat alphabetical order

philandstuff committed:
fix typo

philandstuff commented:
I did one last test run through the instructions and they worked for me 👍
